icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,462 @@
+    }
+    if (b4 != UnicodeString(INVALID) && b5 != UnicodeString(INVALID) && b6 != UnicodeString(INVALID)) {
+        featureList->adoptElement(concatUnicodeString("TB4:", UnicodeString(b4).append(b5).append(b6), status), status);
+    }
+
+    int32_t score = negativeSum;
+    for (int32_t j = 0; j < featureList->size(); j++) {
+        UnicodeString* key = (UnicodeString*)featureList->elementAt(j);
+        if(fModel.containsKey(*key)) {
+            score += (2 * fModel.geti(*key));
+        }
+    }
+
+    if (score > 0) {
+        boundary.addElement(index, status);
+        numBreaks++;
+    }
+
+    featureList->removeAllElements();
+    delete featureList;
+}
+
+void CjMLBreakEngine::initElementList(UnicodeString& inString, UVector* elementList, UErrorCode& status) const {
+    UChar32 w1, w2, w3, w4, w5, w6;
+    UnicodeString b1, b2, b3, b4, b5, b6;
+    w1 = INVALID;
+    w2 = INVALID;
+    w3 = inString.char32At(0);
+    w4 = inString.char32At(1);
+    w5 = 2 < inString.length() ? inString.char32At(2) : INVALID;
+    w6 = 3 < inString.length() ? inString.char32At(3) : INVALID;
+
+    b1 = UnicodeString(INVALID);
+    b2 = UnicodeString(INVALID);
+    b3 = getUnicodeBlockIndex(w3, status);
+    b4 = getUnicodeBlockIndex(w4, status);
+    b5 = getUnicodeBlockIndex(w5, status);
+    b6 = getUnicodeBlockIndex(w6, status);
+
+    elementList->addElement(new Element(w1, b1), status);
+    elementList->addElement(new Element(w2, b2), status);
+    elementList->addElement(new Element(w3, b3), status);
+    elementList->addElement(new Element(w4, b4), status);
+    elementList->addElement(new Element(w5, b5), status);
+    elementList->addElement(new Element(w6, b6), status);
+}
+
+/**
+ * Convert an integer, positive or negative, to a character string radix 10.
+ */
+static char* itoa(int32_t i, char* buf) {
+    char* result = buf;
+    // Handle negative
+    if (i < 0) {
+        *buf++ = '-';
+        i = -i;
+    }
+    // Output digits in reverse order
+    char* p = buf;
+    do {
+        *p++ = (char)('0' + (i % 10));
+        i /= 10;
+    } while (i);
+    *p-- = 0;
+    // Reverse the string
+    while (buf < p) {
+        char c = *buf;
+        *buf++ = *p;
+        *p-- = c;
+    }
+
+    return result;
+}
+
+UnicodeString CjMLBreakEngine::getUnicodeBlockIndex(UChar32 ch, UErrorCode& status) const{
+    U_ASSERT(fUnicodeBlock->size() > 0);
+    if (U_FAILURE(status)) {
+        return UnicodeString();
+    }
+    if (ch == INVALID) {
+        return UnicodeString(INVALID);
+    }
+    int32_t start = 0;
+    int32_t end = fUnicodeBlock->size();
+    while (start < end) {
+        int32_t mid = (start + end) / 2;
+        if (ch < fUnicodeBlock->elementAti(mid)) {
+            end = mid;
+        }
+        else {
+            start = mid + 1;
+        }
+    }
+    //Add '0' to fill up three digits
+    char buff[32];
+    itoa(start, buff);
+    UnicodeString result(buff);
```

Comment:
These three lines should be changed to 
UnicodeString result;
result = ICU_Utility::appendNumber(result, start);

Reply:
Done